File: phases/10-databases-and-storage/08-indexing-hash-bitmap-gist/code/main.py
"""
Indexing — Hash, Bitmap, GiST
Phase 10 — Databases & Storage Systems

Implements:
- ExtendibleHashIndex: directory-based extendible hashing with global/local depth
- BitmapIndex: roaring-bitmap-like index supporting AND/OR queries
- Performance comparison demo
"""

import sys
import time
from typing import Any, Optional


class Bucket:
    def __init__(self, depth: int):
        self.depth = depth
        self.keys: list[int] = []
        self.values: list[int] = []

    @property
    def capacity(self) -> int:
        return 2

    @property
    def is_full(self) -> bool:
        return len(self.keys) >= self.capacity

    def search(self, key: int) -> Optional[int]:
        for i, k in enumerate(self.keys):
            if k == key:
                return self.values[i]
        return None

    def insert(self, key: int, value: int) -> bool:
        for i, k in enumerate(self.keys):
            if k == key:
                self.values[i] = value
                return False
        if self.is_full:
            return False
        self.keys.append(key)
        self.values.append(value)
        return True

    def remove(self, key: int) -> bool:
        for i, k in enumerate(self.keys):
            if k == key:
                self.keys.pop(i)
                self.values.pop(i)
                return True
        return False

    def items(self):
        return zip(self.keys, self.values)


class ExtendibleHashIndex:
    def __init__(self):
        self._global_depth = 1
        self._directory: list[Bucket] = [Bucket(self._global_depth)
                                         for _ in range(1 << self._global_depth)]

    def _hash_int(self, key: int) -> int:
        return hash(key) & 0xFFFFFFFF

    def _mask(self, depth: int) -> int:
        return (1 << depth) - 1

    def _dir_index(self, key: int, depth: int) -> int:
        return self._hash_int(key) & self._mask(depth)

    def search(self, key: int) -> Optional[int]:
        idx = self._dir_index(key, self._global_depth)
        return self._directory[idx].search(key)

    def insert(self, key: int, value: int) -> None:
        idx = self._dir_index(key, self._global_depth)
        bucket = self._directory[idx]
        if not bucket.is_full or bucket.search(key) is not None:
            bucket.insert(key, value)
            return
        self._split(idx, key, value)

    def _split(self, idx: int, key: int, value: int) -> None:
        bucket = self._directory[idx]
        old_depth = bucket.depth
        new_depth = old_depth + 1

        if new_depth > self._global_depth:
            self._double_directory()

        b0 = Bucket(new_depth)
        b1 = Bucket(new_depth)

        mask_bit = 1 << old_depth

        for k, v in bucket.items():
            target = b0 if (self._dir_index(k, new_depth) & mask_bit) == 0 else b1
            target.keys.append(k)
            target.values.append(v)

        target = b0 if (self._dir_index(key, new_depth) & mask_bit) == 0 else b1
        target.keys.append(key)
        target.values.append(value)

        for i, bp in enumerate(self._directory):
            if bp is bucket:
                if (i & mask_bit) == 0:
                    self._directory[i] = b0
                else:
                    self._directory[i] = b1

    def _double_directory(self) -> None:
        old_size = len(self._directory)
        self._directory.extend(self._directory[i % old_size]
                               for i in range(old_size))
        self._global_depth += 1

    def remove(self, key: int) -> bool:
        idx = self._dir_index(key, self._global_depth)
        return self._directory[idx].remove(key)


class Bitmap:
    def __init__(self, size: int = 0):
        self.size = size
        nwords = (size + 63) // 64 if size else 0
        self.words: list[int] = [0] * nwords

    def set(self, pos: int) -> None:
        while pos >= self.size:
            self.size = max(self.size * 2, 64)
            needed = (self.size + 63) // 64
            self.words.extend([0] * (needed - len(self.words)))
        self.words[pos // 64] |= 1 << (pos % 64)

    def get(self, pos: int) -> int:
        if pos >= self.size:
            return 0
        return (self.words[pos // 64] >> (pos % 64)) & 1

    def __and__(self, other: "Bitmap") -> "Bitmap":
        n = min(len(self.words), len(other.words))
        r = Bitmap()
        r.words = [self.words[i] & other.words[i] for i in range(n)]
        r.size = min(self.size, other.size)
        return r

    def __or__(self, other: "Bitmap") -> "Bitmap":
        n = max(len(self.words), len(other.words))
        r = Bitmap()
        r.words = [0] * n
        for i in range(n):
            a = self.words[i] if i < len(self.words) else 0
            b = other.words[i] if i < len(other.words) else 0
            r.words[i] = a | b
        r.size = max(self.size, other.size)
        return r

    def __invert__(self) -> "Bitmap":
        r = Bitmap()
        r.size = self.size
        mask = (1 << 64) - 1
        r.words = [~w & mask for w in self.words]
        return r

    def count(self) -> int:
        return sum(w.bit_count() for w in self.words)


class RoaringBitmap:
    CHUNK_SHIFT = 16
    CHUNK_SIZE = 1 << CHUNK_SHIFT
    SPARSE_THRESHOLD = 4096

    class Chunk:
        def __init__(self):
            self.dense: Optional[Bitmap] = None
            self.sparse: list[int] = []

    def __init__(self):
        self.chunks: dict[int, "RoaringBitmap.Chunk"] = {}

    def _chunk(self, key: int) -> "RoaringBitmap.Chunk":
        hi = key >> self.CHUNK_SHIFT
        if hi not in self.chunks:
            self.chunks[hi] = RoaringBitmap.Chunk()
        return self.chunks[hi]

    def add(self, key: int) -> None:
        lo = key & (self.CHUNK_SIZE - 1)
        c = self._chunk(key)
        if c.dense is not None:
            c.dense.set(lo)
        else:
            if lo not in c.sparse:
                c.sparse.append(lo)
            if len(c.sparse) > self.SPARSE_THRESHOLD:
                bm = Bitmap(self.CHUNK_SIZE)
                for v in c.sparse:
                    bm.set(v)
                c.dense = bm
                c.sparse = []

    def contains(self, key: int) -> bool:
        hi = key >> self.CHUNK_SHIFT
        if hi not in self.chunks:
            return False
        lo = key & (self.CHUNK_SIZE - 1)
        c = self.chunks[hi]
        if c.dense is not None:
            return c.dense.get(lo) == 1
        return lo in c.sparse

    def __and__(self, other: "RoaringBitmap") -> "RoaringBitmap":
        r = RoaringBitmap()
        for hi, c in self.chunks.items():
            if hi not in other.chunks:
                continue
            oc = other.chunks[hi]
            if c.dense is not None and oc.dense is not None:
                rc = RoaringBitmap.Chunk()
                rc.dense = c.dense & oc.dense
                r.chunks[hi] = rc
            elif c.dense is None and oc.dense is None:
                inter = sorted(set(c.sparse) & set(oc.sparse))
                if inter:
                    rc = RoaringBitmap.Chunk()
                    rc.sparse = inter
                    r.chunks[hi] = rc
            else:
                dense = c.dense if c.dense is not None else oc.dense
                sparse = oc.sparse if c.dense is not None else c.sparse
                rc = RoaringBitmap.Chunk()
                rc.sparse = [v for v in sparse if dense.get(v)]
                r.chunks[hi] = rc
        return r

    def __or__(self, other: "RoaringBitmap") -> "RoaringBitmap":
        r = RoaringBitmap()
        for hi, c in self.chunks.items():
            r.chunks[hi] = c
        for hi, c in other.chunks.items():
            if hi in r.chunks:
                ex = r.chunks[hi]
                if c.dense is not None and ex.dense is not None:
                    ex.dense = ex.dense | c.dense
                elif c.dense is None and ex.dense is None:
                    ex.sparse = sorted(set(ex.sparse) | set(c.sparse))
                    if len(ex.sparse) > self.SPARSE_THRESHOLD:
                        bm = Bitmap(self.CHUNK_SIZE)
                        for v in ex.sparse:
                            bm.set(v)
                        ex.dense = bm
                        ex.sparse = []
                elif c.dense is not None:
                    for v in ex.sparse:
                        c.dense.set(v)
                    ex.dense = c.dense
                    ex.sparse = []
                else:
                    for v in c.sparse:
                        if ex.dense.get(v) == 0:
                            ex.sparse.append(v)
                    ex.sparse = sorted(set(ex.sparse))
                    if len(ex.sparse) > self.SPARSE_THRESHOLD:
                        bm = Bitmap(self.CHUNK_SIZE)
                        for v in ex.sparse:
                            bm.set(v)
                        ex.dense = bm
                        ex.sparse = []
            else:
                r.chunks[hi] = c
        return r

    def cardinality(self) -> int:
        total = 0
        for c in self.chunks.values():
            if c.dense is not None:
                total += c.dense.count()
            else:
                total += len(c.sparse)
        return total


class BitmapIndex:
    def __init__(self, column_name: str):
        self.column_name = column_name
        self.bitmaps: dict[Any, RoaringBitmap] = {}

    def insert(self, row_id: int, value: Any) -> None:
        if value not in self.bitmaps:
            self.bitmaps[value] = RoaringBitmap()
        self.bitmaps[value].add(row_id)

    def query_eq(self, value: Any) -> RoaringBitmap:
        return self.bitmaps.get(value, RoaringBitmap())

    def query_in(self, values: list[Any]) -> RoaringBitmap:
        r = RoaringBitmap()
        for v in values:
            if v in self.bitmaps:
                r = r | self.bitmaps[v]
        return r

    def query_and(self, v1: Any, v2: Any) -> RoaringBitmap:
        b1 = self.bitmaps.get(v1, RoaringBitmap())
        b2 = self.bitmaps.get(v2, RoaringBitmap())
        return b1 & b2


def demo_extendible_hash():
    print("=== Extendible Hash Index Demo ===")
    idx = ExtendibleHashIndex()
    for k, v in [(10, 100), (22, 200), (1, 10), (7, 70),
                 (15, 150), (3, 30), (31, 310), (9, 90)]:
        idx.insert(k, v)
    print(f"global_depth={idx._global_depth}, dir_size={len(idx._directory)}")
    for k in [10, 22, 1, 7, 15, 3, 31, 9, 99]:
        v = idx.search(k)
        print(f"  search({k:3d}) -> {v}")
    print()


def demo_bitmap_index():
    print("=== Bitmap Index Demo ===")
    idx = BitmapIndex("status")
    statuses = ["pending", "shipped", "cancelled", "shipped", "pending",
                "shipped", "pending", "cancelled", "shipped", "shipped"]
    for row_id, s in enumerate(statuses):
        idx.insert(row_id, s)
    shipped = idx.query_eq("shipped")
    cancelled = idx.query_eq("cancelled")
    print(f"  shipped:   {[i for i in range(10) if shipped.contains(i)]}")
    print(f"  cancelled: {[i for i in range(10) if cancelled.contains(i)]}")
    combined = shipped | cancelled
    print(f"  shipped|cancelled: {[i for i in range(10) if combined.contains(i)]}")

    idx2 = BitmapIndex("region")
    for row_id, r in enumerate([1, 2, 1, 3, 2, 1, 3, 2, 1, 2]):
        idx2.insert(row_id, r)
    region1 = idx2.query_eq(1)
    region2 = idx2.query_eq(2)
    shipped_region1 = shipped & region1
    print(f"  shipped & region=1: {[i for i in range(10) if shipped_region1.contains(i)]}")
    shipped_or_cancelled_r2 = (shipped | cancelled) & region2
    print(f"  (shipped|cancelled) & region=2: {[i for i in range(10) if shipped_or_cancelled_r2.contains(i)]}")
    print()


def demo_roaring():
    print("=== Roaring Bitmap Compression Demo ===")
    rb1 = RoaringBitmap()
    for i in range(0, 100000, 3):
        rb1.add(i)
    rb2 = RoaringBitmap()
    for i in range(0, 100000, 5):
        rb2.add(i)
    r_and = rb1 & rb2
    r_or = rb1 | rb2
    print(f"  multiples of 3: {rb1.cardinality()}")
    print(f"  multiples of 5: {rb2.cardinality()}")
    print(f"  multiples of 15: {r_and.cardinality()}")
    print(f"  multiples of 3 or 5: {r_or.cardinality()}")
    print()


def demo_performance():
    print("=== Performance Comparison ===")
    n = 50000

    t0 = time.perf_counter()
    eh = ExtendibleHashIndex()
    for i in range(n):
        eh.insert(i, i * 10)
    for i in range(n):
        eh.search(i)
    t_eh = time.perf_counter() - t0

    t0 = time.perf_counter()
    bm = BitmapIndex("even")
    for i in range(n):
        bm.insert(i, i % 2)
    for i in range(n):
        bm.query_eq(i % 2)
    t_bm = time.perf_counter() - t0

    print(f"  ExtendibleHash: {n} inserts + {n} searches: {t_eh:.4f}s")
    print(f"  BitmapIndex:    {n} inserts + {n} queries:    {t_bm:.4f}s")
    print()


def main():
    demo_extendible_hash()
    demo_bitmap_index()
    demo_roaring()
    demo_performance()


if __name__ == "__main__":
    main()

File: phases/10-databases-and-storage/08-indexing-hash-bitmap-gist/code/main.rs
//! Indexing — Hash, Bitmap, GiST
//! Phase 10 — Databases & Storage Systems
//!
//! Implementes extendible hashing with directory doubling and bucket splitting.

use std::fmt::Debug;
use std::hash::{Hash, Hasher};

// ---------------------------------------------------------------------------
// Bucket
// ---------------------------------------------------------------------------

#[derive(Debug, Clone)]
struct Bucket<K: Clone + Eq + Hash, V: Clone> {
    depth: u32,
    keys: Vec<K>,
    values: Vec<V>,
}

impl<K: Clone + Eq + Hash, V: Clone> Bucket<K, V> {
    fn new(depth: u32) -> Self {
        Bucket { depth, keys: Vec::new(), values: Vec::new() }
    }

    fn capacity(&self) -> usize {
        2
    }

    fn is_full(&self) -> bool {
        self.keys.len() >= self.capacity()
    }

    fn search(&self, key: &K) -> Option<&V> {
        self.keys.iter().position(|k| k == key).map(|i| &self.values[i])
    }

    fn upsert(&mut self, key: K, value: V) -> bool {
        if let Some(i) = self.keys.iter().position(|k| *k == key) {
            self.values[i] = value;
            return false;
        }
        if self.is_full() {
            return false;
        }
        self.keys.push(key);
        self.values.push(value);
        true
    }

    fn remove(&mut self, key: &K) -> bool {
        if let Some(i) = self.keys.iter().position(|k| k == key) {
            self.keys.swap_remove(i);
            self.values.swap_remove(i);
            return true;
        }
        false
    }
}

// ---------------------------------------------------------------------------
// ExtendibleHash
// ---------------------------------------------------------------------------

type BucketPtr<K, V> = *mut Bucket<K, V>;

pub struct ExtendibleHash<K: Clone + Eq + Hash, V: Clone> {
    global_depth: u32,
    directory: Vec<BucketPtr<K, V>>,
}

unsafe impl<K: Clone + Eq + Hash + Send, V: Clone + Send> Send for ExtendibleHash<K, V> {}
unsafe impl<K: Clone + Eq + Hash + Sync, V: Clone + Sync> Sync for ExtendibleHash<K, V> {}

impl<K: Clone + Eq + Hash, V: Clone> ExtendibleHash<K, V> {
    pub fn new() -> Self {
        let b = Box::into_raw(Box::new(Bucket::new(1)));
        ExtendibleHash { global_depth: 1, directory: vec![b, b] }
    }

    fn hash_key(key: &K) -> u64 {
        let mut h = std::hash::DefaultHasher::new();
        key.hash(&mut h);
        h.finish()
    }

    fn mask(depth: u32) -> u32 {
        (1 << depth) - 1
    }

    fn dir_index(hash: u64, depth: u32) -> usize {
        (hash as u32 & Self::mask(depth)) as usize
    }

    pub fn search(&self, key: &K) -> Option<&V> {
        let h = Self::hash_key(key);
        let idx = Self::dir_index(h, self.global_depth);
        unsafe { (*self.directory[idx]).search(key) }
    }

    pub fn insert(&mut self, key: K, value: V) {
        let h = Self::hash_key(&key);
        let idx = Self::dir_index(h, self.global_depth);
        let bp = self.directory[idx];
        unsafe {
            if !(*bp).is_full() || (*bp).search(&key).is_some() {
                (*bp).upsert(key, value);
                return;
            }
        }
        self.split(idx, key, value, h);
    }

    fn split(&mut self, idx: usize, key: K, value: V, hash: u64) {
        let old_depth;
        unsafe { old_depth = (*self.directory[idx]).depth; }
        let new_depth = old_depth + 1;

        if new_depth > self.global_depth {
            self.double_directory();
        }

        let mut b0 = Box::new(Bucket::new(new_depth));
        let mut b1 = Box::new(Bucket::new(new_depth));

        let mask_bit = 1 << old_depth;

        unsafe {
            let old = &mut *self.directory[idx];
            let keys = std::mem::take(&mut old.keys);
            let values = std::mem::take(&mut old.values);
            for (k, v) in keys.into_iter().zip(values.into_iter()) {
                let hk = Self::hash_key(&k);
                if (Self::dir_index(hk, new_depth) & mask_bit) == 0 {
                    b0.keys.push(k);
                    b0.values.push(v);
                } else {
                    b1.keys.push(k);
                    b1.values.push(v);
                }
            }
        }

        if (Self::dir_index(hash, new_depth) & mask_bit) == 0 {
            b0.keys.push(key);
            b0.values.push(value);
        } else {
            b1.keys.push(key);
            b1.values.push(value);
        }

        let bp0 = Box::into_raw(b0);
        let bp1 = Box::into_raw(b1);

        let old_bucket_ptr = self.directory[idx];
        for i in 0..self.directory.len() {
            if self.directory[i] == old_bucket_ptr {
                if (i as u32 & mask_bit as u32) == 0 {
                    self.directory[i] = bp0;
                } else {
                    self.directory[i] = bp1;
                }
            }
        }
    }

    fn double_directory(&mut self) {
        let old = self.directory.clone();
        self.directory.reserve(old.len());
        for i in 0..old.len() {
            self.directory.push(old[i]);
        }
        self.global_depth += 1;
    }

    pub fn remove(&mut self, key: &K) -> bool {
        let h = Self::hash_key(key);
        let idx = Self::dir_index(h, self.global_depth);
        unsafe { (*self.directory[idx]).remove(key) }
    }
}

impl<K: Clone + Eq + Hash, V: Clone> Drop for ExtendibleHash<K, V> {
    fn drop(&mut self) {
        let mut freed = std::collections::HashSet::new();
        for &p in &self.directory {
            if freed.insert(p as usize) {
                unsafe { drop(Box::from_raw(p)); }
            }
        }
    }
}

// ---------------------------------------------------------------------------
// Demo
// ---------------------------------------------------------------------------

fn demo_extendible_hash() {
    let mut idx = ExtendibleHash::new();
    let pairs = [(10, 100), (22, 200), (1, 10), (7, 70),
                 (15, 150), (3, 30), (31, 310), (9, 90)];

    for &(k, v) in &pairs {
        idx.insert(k, v);
    }

    println!("global_depth={}", idx.global_depth);
    println!("directory_size={}", idx.directory.len());

    let tests = [10, 22, 1, 7, 15, 3, 31, 9, 99];
    for &k in &tests {
        match idx.search(&k) {
            Some(v) => println!("  search({:3}) -> {}", k, v),
            None => println!("  search({:3}) -> not found", k),
        }
    }
}

fn main() {
    println!("=== Extendible Hash Index (Rust) ===");
    demo_extendible_hash();
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_insert_and_search() {
        let mut idx = ExtendibleHash::new();
        idx.insert(1, 10);
        idx.insert(2, 20);
        idx.insert(3, 30);
        assert_eq!(idx.search(&1), Some(&10));
        assert_eq!(idx.search(&2), Some(&20));
        assert_eq!(idx.search(&3), Some(&30));
        assert_eq!(idx.search(&4), None);
    }

    #[test]
    fn test_update_existing() {
        let mut idx = ExtendibleHash::new();
        idx.insert(5, 50);
        idx.insert(5, 500);
        assert_eq!(idx.search(&5), Some(&500));
    }

    #[test]
    fn test_remove() {
        let mut idx = ExtendibleHash::new();
        idx.insert(42, 420);
        assert!(idx.search(&42).is_some());
        assert!(idx.remove(&42));
        assert!(idx.search(&42).is_none());
        assert!(!idx.remove(&42));
    }

    #[test]
    fn test_directory_doubling() {
        let mut idx = ExtendibleHash::new();
        for i in 0..100 {
            idx.insert(i, i * 10);
        }
        for i in 0..100 {
            assert_eq!(idx.search(&i), Some(&(i * 10)));
        }
    }

    #[test]
    fn test_string_keys() {
        let mut idx = ExtendibleHash::new();
        idx.insert("hello".to_string(), 1);
        idx.insert("world".to_string(), 2);
        assert_eq!(idx.search(&"hello".to_string()), Some(&1));
        assert_eq!(idx.search(&"world".to_string()), Some(&2));
    }
}
